README.md
# class-project-Rust-bag

A Bag as an unsorted, ordered collection of generic types in which the same item may appear multiple times. The bag's capacity can grow as needed and can be reduced.

This was a class project to construct a simple indepth data structure in Rust. This is one of the first Rust programs I have ever written as well as the longest.

# Author
Nathan Bradley based on the java implementation from H. Paul Haiduk with credit given to Michael Main


changelog.txt
Implemented:
    4.4.2018
    Assess functions in bag for a different implementation.
        occurrences()
            Similar to the implementation observed from another student.
        Debug::fmt()
            All for nawt until I found the for_each() iterator function.
        Add::add()
            Couldnt use iterators but the code is nicer none the less.
        PartialEq::eq()
            I used the built in slice eq trait, making it much simpler.
        trim_to_size
            A small expression change.
        erase_one
            Used iterators and pattern matching.
        varius methods for reference parameters
            Parameters that are copyable, like usize, were left as non references.

    Added Clone::clone_from() method for completeness

    
    5.4.2018
    Restructured bagexam.rs to fit more of a testing implementation.
        It is now five tests with no print statements.
        I added a few more comparisons into it.

    Made a f64 bag
        Very simple change of a few parameters.

    8.4.2018
    Removed reference parameters for matching standard Rust convention.

    Added a 3rd-party library of traits that matches the methods that were implemented.
        The most visible change is the names of the methods. The implementation is the same.
        This sets the naminging to match the convention seen in Rust.
        Gives me experience in using a library that is not in the std.

    Added documentation for most files in the project:
        bagexamf64.rs, bagexam.rs, bag.rs, bagf64.rs, bagVec.rs
        Included verbose descriptions and serveral examples.

    14.4.2018
    Finished (probably) documentation for all files.
        Added comments for which methods matched the expressed implemented methods.

    Added a forward iterator
        The iterator involved a bag reference, which both ment no modification with
        an active iterator and also explicit lifetime values. Added a few other
        iterator funtions with the forward stepping.

    Implemented type aliase for non-generic bag
        If I had read carfully I would have realized earlier to not
        implement this for the non-generic bag.

In_Combat:
    <None>

Planned:
    <None>

Abandonded:
    Assess functions in bag for a different implementation
        AddAssign::add_assign()
            I considered using what I used in Add::add()
                but I believe my existing code is more efficient.
        erase
            Not only a problem in efficientcy, but also a problem of iterative implementation.
        insert
            No obvious reason for a more complicated solution.


src/bagtest.rs
// FILE: bagtest.rs

//! An interactive test program for the bag generic structure.
//!
//! Run with `cargo run`
//!
//! # The following commands are available with 2 bags
//!
//! * `A` Use the assignment operator to make bag1 equal to bag2
//!
//! * `a` Use the assignment operator to make bag2 equal to bag1
//!
//! * `C` Use the copy constructor to make bag1 equal to bag2
//!
//! * `c` Use the copy constructor to make bag2 equal to bag1
//!
//! * `I` Insert an item into bag1
//!
//! * `i` Insert an item into bag2
//!
//! * `R` Erase all of item from bag1
//!
//! * `r` Erase all of item from bag2
//!
//! * `X` Erase one of item from bag1
//!
//! * `x` Erase one of item from bag2
//!
//! * `O` Display both bags
//!
//! * `S` Print the result from the len( ) functions
//!
//! * `Q` Quit this test program

use bag::Bag;

#[ allow( unused_imports ) ]
use bag::len_trait::{Len, Capacity, CapacityMut};
use std::io::Write;
use std::ascii::AsciiExt;

macro_rules! read_input
{
    () =>
    {{
        ::std::io::stdout().flush().unwrap();
        let mut input = String::new();
        ::std::io::stdin().read_line( &mut input ).expect( "Error: could not read input" );
        input
    }};
}

pub fn bagtest()
{
    // bags that we'll perform tests on
    let mut bag1: Bag<f32> = Bag::new();
    let mut bag2: Bag<f32> = Bag::new();

    let mut choice; // A command character entered by the user

    println!( "\tI have initialized two empty bags of doubles." );


    while
    {
        print_menu();
        choice = get_user_command();

        match choice
        {
            'A' => { bag1 = copybag( &bag2 ); }
            'a' => { bag2 = copybag( &bag1 ); }
            'C' => { bag1.clone_from( &bag2 ); }
            'c' => { bag2.clone_from( &bag1 ); }
            'S' | 's' => { println!( "The bags' sizes are {} and {}",
                                        bag1.len(), bag2.len() ); }
            'I' => { bag1.insert( get_number() ); }
            'i' => { bag2.insert( get_number() ); }
            'R' => { bag1.erase( get_number() ); }
            'r' => { bag2.erase( get_number() ); }
            'X' => { bag1.erase_one( get_number() ); }
            'x' => { bag2.erase_one( get_number() ); }
            'O' | 'o' => { show_bags( &bag1, &bag2 ); }
            'q' | 'Q' => { println!( "Ridicule is the best test of truth." ); }
            val => { println!( "{} is invalid. Sorry.", val ); }
        }

        !choice.eq_ignore_ascii_case( &'Q' ) // post check
    }{}
}

// A menu of choices for this program has been written to cout.
fn print_menu()
{
    println!( "\n\n\tThe following choices are available with 2 bags: " );
    println!( "\t\t A  Use the assignment operator to make bag1 equal to bag2" );
    println!( "\t\t a  Use the assignment operator to make bag2 equal to bag1" );
    println!( "\t\t C  Use the copy constructor to make bag1 equal to bag2" );
    println!( "\t\t c  Use the copy constructor to make bag2 equal to bag1" );
    println!( "\t\t I  Insert an item into bag1" );
    println!( "\t\t i  Insert an item into bag2" );
    println!( "\t\t R  Erase all of item from bag1" );
    println!( "\t\t r  Erase all of item from bag2" );
    println!( "\t\t X  Erase one of item from bag1" );
    println!( "\t\t x  Erase one of item from bag2" );
    println!( "\t\t O  Display both bags" );
    println!( "\t\t S  Print the result from the len( ) functions" );
    println!( "\t\t Q  Quit this test program" );
}

// The user has been prompted to enter a one character command.
// A line of input (with at least one character) has been read,
// and the first character of the input line is returned.
fn get_user_command() -> char
{
    print!( "\t\tEnter choice: " );

    read_input!().chars().next().expect( "Error: could not parse character" )
}

// The function has tested whether the numbers 0..9 are in
// the two bags, and printed the results to standard output.
fn show_bags( bag1: &Bag<f32>, bag2: &Bag<f32> )
{
    println!( "\t\tbag1 {:?}\n\t\tbag2 {:?}",
                bag1, bag2 );
}

// The return value is a copy of b.
fn copybag( input_bag: &Bag<f32> ) -> Bag<f32>
{
    input_bag.clone()
}

// The user has been prompted to enter bag::value_type  number. The number
// has been read, echoed to the screen, and returned by the function.
fn get_number() -> f32
{
    print!( "\t\tPlease enter a bag::value_type number for the bag: ");
    let result: f32 = read_input!().trim_right_matches( '\n' ).parse::<f32>().expect( "Error: could not parse f32" );
    println!( "\t\t{} has been read.", result );
    result
}


src/bagexamf64.rs
// FILE: bagexamf64.rs

//! This module holds a variety of tests for the bagf64 structure
//!
//! 1) basic tests of insert and the constant member functions
//!
//! 2) tests of the copy constructor and == method
//!
//! 3) tests of the assignment operator
//!
//! 4) basic tests for the erase functions
//!
//! 5) basic tests for the += and + functions
//!
//! 6) tests for the bag iterator
//!
//! Run with `cargo test`

extern crate rand;

use bagf64::Bag;
use bag::len_trait::{Len, Capacity, CapacityMut};

#[ test ]
///   Performs some basic tests of insert and the constant member functions. 
fn test1()
{
    #[ allow( non_snake_case ) ]
    let TEST_SIZE = 3000;
    let mut test_bag = Bag::new();

    // A. Testing size for an empty bag.
    assert_eq!( test_bag.len(), 0 );

    // B. Adding the number 4 to the bag, and then testing size.
    test_bag.insert( 4.0 );
    assert_eq!( test_bag.len(), 1 );

    // C. Inserting the number 2 into the bag. Then checking size
    test_bag.insert( 2.0 );
    assert_eq!( test_bag.len(), 2 );

    // D. Inserting the number 1 into the bag. Then checking size.
    test_bag.insert( 1.0 );
    assert_eq!( test_bag.len(), 3 );

    // E. Inserting the number 3 into the bag. Then checking size.
    test_bag.insert( 3.0 );
    assert_eq!( test_bag.len(), 4 );

    // F. Inserting another 2 into the bag. Then checking size.
    test_bag.insert( 2.0 );
    assert_eq!( test_bag.len(), 5 );

    //    Then checking occurrences of 2.
    assert_eq!( test_bag.occurrences( 2.0 ), 2 );

    // G. Inserting the numbers 5, 6, and 7 into the bag. Then checking size.
    test_bag.insert( 5.0 );
    test_bag.insert( 6.0 );
    test_bag.insert( 7.0 );
    assert_eq!( test_bag.len(), 8 );

    // H. Inserting two more 2's into the bag. And then checking occurrences of 2's.
    test_bag.insert( 2.0 );
    test_bag.insert( 2.0 );
    assert_eq!( test_bag.occurrences( 2.0 ), 4 );

    // I. Inserting TEST_SIZE random items between 0 and 49 and then checking size.

    for _ in 0..TEST_SIZE
    {
        test_bag.insert( ( rand::random::<f64>() % 50.0 ) );
    }

    assert_eq!( test_bag.len(), TEST_SIZE + 10 );
}

#[ test ]
///   Performs some tests of the copy constructor and == method.
fn test2()
{
    let mut test_bag = Bag::new();

    // A. Testing that copy constructor works okay for empty bag...
    let copy1 = test_bag.clone();
    assert_eq!( copy1.len(), 0 );

    // B. Testing copy constructor with 4-item bag...
    test_bag.insert( 1.0 );
    test_bag.insert( 1.0 );
    test_bag.insert( 1.0 );
    test_bag.insert( 1.0 );
    let copy2 = test_bag.clone();

    //    and now testing the == method...
    assert_eq!( test_bag, copy2 );
    assert_eq!( copy2, test_bag );

    test_bag.insert( 1.0 ); // Alter the original, but not the copy
    // C. Then checking size of copy.
    assert_eq!( copy2.len(), 4 );

    // D. Altering original but not the copy.
    assert_eq!( test_bag.len(), 5 );
}

#[ test ]
///   Performs some tests of the assignment operator.
fn test3()
{
    let mut test_bag = Bag::new();

    // A. Testing that assignment operator works okay for empty bag...
    let mut copy1 = Bag::new();
    copy1.insert( 1.0 );
    copy1 = test_bag.clone();

    assert_eq!( copy1.len(), 0 );

    // B. Testing assignment operator with 4-item bag...
    test_bag.insert( 1.0 );
    test_bag.insert( 1.0 );
    test_bag.insert( 1.0 );
    test_bag.insert( 1.0 );
    let mut copy2 = Bag::new();
    copy2.clone_from( &test_bag );
    test_bag.insert( 1.0 ); // Alter the original, but not the copy
    //    altering original by an insertion...

    assert_eq!( test_bag.occurrences( 1.0 ), 5 );
    assert_eq!( copy2.occurrences( 1.0 ), 4 );

    // Test passed.\n   testing size of assigned to...
    assert_eq!( copy2.len(), 4 );

    //    testing size of original...
    assert_eq!( test_bag.len(), 5 );
}

#[ test ]
///   Performs basic tests for the erase functions
fn test4()
{
    let mut test_bag = Bag::new();

    // Testing erase from empty bag, should have no effect.
    assert_eq!( test_bag.erase( 0.0 ), 0 );
    assert_eq!( test_bag.len(), 0 );
        
    // Inserting these: 8 6 10 1 7 10 15 3 13 2 5 11 14 4 12
    test_bag.insert(  8.0 );
    test_bag.insert(  6.0 );
    test_bag.insert( 10.0 );
    test_bag.insert(  1.0 );
    test_bag.insert(  7.0 );
    test_bag.insert( 10.0 );
    test_bag.insert( 15.0 );
    test_bag.insert(  3.0 );
    test_bag.insert( 13.0 );
    test_bag.insert(  2.0 );
    test_bag.insert(  5.0 );
    test_bag.insert( 11.0 );
    test_bag.insert( 14.0 );
    test_bag.insert(  4.0 );
    test_bag.insert( 12.0 );
    assert_eq!( test_bag.len(), 15 );

    // Now testing capacity, should be 16.
    assert_eq!( test_bag.capacity(), 16 );

    // Erasing 0, which is not in bag, so bag should be unchanged.
    assert!( !test_bag.erase_one( 0.0 ) );
    assert_eq!( test_bag.len(), 15 );

    // Erasing the 6.
    assert_eq!( test_bag.erase( 6.0 ), 1 );
    assert_eq!( test_bag.len(), 14 );

    // Erasing one 10.
    assert!( test_bag.erase_one( 10.0 ) );
    assert_eq!( test_bag.len(), 13 );

    // Erasing the 1.
    assert_eq!( test_bag.erase( 1.0 ), 1 );
    assert_eq!( test_bag.len(), 12 );

    // Erasing the 15.
    assert_eq!( test_bag.erase( 15.0 ), 1 );
    assert_eq!( test_bag.len(), 11 );

    // Erasing the 5.
    assert_eq!( test_bag.erase( 5.0 ), 1 );
    assert_eq!( test_bag.len(), 10 );

    // Erasing the 11.
    assert_eq!( test_bag.erase( 11.0 ), 1 );
    assert_eq!( test_bag.len(), 9 );

    // Erasing the 3.
    assert_eq!( test_bag.erase( 3.0 ), 1 );
    assert_eq!( test_bag.len(), 8 );

    // Erasing the 13.
    assert_eq!( test_bag.erase( 13.0 ), 1 );
    assert_eq!( test_bag.len(), 7 );

    // Erasing the 2.
    assert_eq!( test_bag.erase( 2.0 ), 1 );
    assert_eq!( test_bag.len(), 6 );

    // Erasing the one and only 14.
    assert!( test_bag.erase_one( 14.0 ) );
    assert_eq!( test_bag.len(), 5 );

    // Erasing the 4.
    assert_eq!( test_bag.erase( 4.0 ), 1 );
    assert_eq!( test_bag.len(), 4 );

    // Erasing the 12.
    assert_eq!( test_bag.erase( 12.0 ), 1 );
    assert_eq!( test_bag.len(), 3 );

    // Erasing the 8.
    assert_eq!( test_bag.erase( 8.0 ), 1 );
    assert_eq!( test_bag.len(), 2 );

    // Erasing the 7.
    assert_eq!( test_bag.erase( 7.0 ), 1 );
    assert_eq!( test_bag.len(), 1 );

    // Erasing the other 10.
    assert!( test_bag.erase_one( 10.0 ) );
    assert_eq!( test_bag.len(), 0 );

    // Testing capacity again.
    assert_eq!( test_bag.capacity(), 16 );

    // Now trimming to size.
    test_bag.shrink_to_fit();
    assert_eq!( test_bag.capacity(), 1 );

    // Now trimming to size again.
    test_bag.shrink_to_fit();
    assert_eq!( test_bag.capacity(), 1 );

    // Inserting value 5000 into the bag.
    // Inserting three 5's into the bag and then erasing all of them.
    test_bag.insert( 5000.0 );
    test_bag.insert( 5.0 ); 
    test_bag.insert( 5.0 );
    test_bag.insert( 5.0 );
    assert_eq!( test_bag.erase( 5.0 ), 3 );
    assert_eq!( test_bag.len(), 1 );
}

#[ test ]
///   Performs basic tests for the += and + functions
fn test5()
{
    let mut test_bag1 = Bag::new();
    let mut test_bag2 = Bag::new();
    let mut test_bag3 = Bag::new();

    // Inserting 2000 1's into test_bag1 and 2000 2's into test_bag2.

    for _ in 0..2000
    {
        test_bag1.insert( 1.0 );
        test_bag2.insert( 2.0 );
    }

    // Now testing the += function.
    test_bag1 += test_bag2.clone();

    //   and now testing for occurrences of 1's and 2's in test_bag1.
    assert_eq!( test_bag1.occurrences( 1.0 ), 2000 ); 
    assert_eq!( test_bag2.occurrences( 2.0 ), 2000 );

    // Now testing the + function.
    test_bag3.clone_from( &( test_bag2.clone() + test_bag2.clone() ) );

    //   and now testing for occurrences of 2's in test_bag3.
    assert_eq!( test_bag3.occurrences( 2.0 ), 4000 );
}

#[ test ]
///   Performs tests for the bag iterator
fn test6()
{
    let mut testbag = Bag::new();

    { // Testing next(), last(), and count()
        let mut iter = testbag.into_iter();
        assert_eq!( iter.next(), None );
        assert_eq!( iter.last(), None );
        assert_eq!( testbag.into_iter().count(), 0 );
    }

    testbag.insert( 8.0 );

    { // Testing next(), last(), and count() again
        let mut iter = testbag.into_iter();
        assert_eq!( iter.next(), Some( &8.0 ) );
        assert_eq!( iter.last(), Some( &8.0 ) );
        assert_eq!( testbag.into_iter().count(), 1 );
    }

    testbag.insert( 3.0 );
    testbag.insert( 4.0 );
    testbag.insert( 7.0 );

    { // Testing next(), last(), and count() again
        let mut iter = testbag.into_iter();
        assert_eq!( iter.next(), Some( &8.0 ) );
        assert_eq!( iter.last(), Some( &7.0 ) );
        assert_eq!( testbag.into_iter().count(), 4 );
    }

    testbag.insert( 2.0 );

    {
        let mut iter = testbag.into_iter();

        // Testing the full range of the bag with next()
        assert_eq!( iter.next(), Some( &8.0 ) );
        assert_eq!( iter.next(), Some( &3.0 ) );
        assert_eq!( iter.next(), Some( &4.0 ) );
        assert_eq!( iter.next(), Some( &7.0 ) );
        assert_eq!( iter.next(), Some( &2.0 ) );

        // ... and then None once it's over.
        assert_eq!( iter.next(), None );

        // More calls may or may not return None. Here, they always will.
        assert_eq!( iter.next(), None );
        assert_eq!( iter.next(), None );
    }
    { // Testing the nth( _ ) method
        let mut iter = testbag.into_iter();

        assert_eq!( iter.nth( 1 ), Some( &3.0 ) );
        assert_eq!( iter.nth( 1 ), Some( &7.0 ) );
        assert_eq!( iter.nth( 1 ), None );
    }
    {
        let mut iter = testbag.into_iter();

        assert_eq!( iter.nth( 0 ), Some( &8.0 ) );
        assert_eq!( iter.nth( 2 ), Some( &7.0 ) );
        assert_eq!( iter.nth( 9 ), None );
    }
}


src/bagexam.rs
// FILE: bagexam.rs

//! This module holds a variety of tests for the bag structure
//!
//! 1) basic tests of insert and the constant member functions
//!
//! 2) tests of the copy constructor and == method
//!
//! 3) tests of the assignment operator
//!
//! 4) basic tests for the erase functions
//!
//! 5) basic tests for the += and + functions
//!
//! 6) tests for the bag iterator
//!
//! Run with `cargo test`

extern crate rand;

use bag::Bag;
//use bagVec::Bag;
use bag::len_trait::{Len, Capacity, CapacityMut};

#[ test ]
///   Performs some basic tests of insert and the constant member functions.
fn test1()
{
    #[ allow( non_snake_case ) ]
    let TEST_SIZE = 3000;
    let mut test_bag = Bag::new();

    // A. Testing size for an empty bag.
    assert_eq!( test_bag.len(), 0 );

    // B. Adding the number 4 to the bag, and then testing size.
    test_bag.insert( 4 );
    assert_eq!( test_bag.len(), 1 );

    // C. Inserting the number 2 into the bag. Then checking size
    test_bag.insert( 2 );
    assert_eq!( test_bag.len(), 2 );

    // D. Inserting the number 1 into the bag. Then checking size.
    test_bag.insert( 1 );
    assert_eq!( test_bag.len(), 3 );

    // E. Inserting the number 3 into the bag. Then checking size.
    test_bag.insert( 3 );
    assert_eq!( test_bag.len(), 4 );

    // F. Inserting another 2 into the bag. Then checking size.
    test_bag.insert( 2 );
    assert_eq!( test_bag.len(), 5 );

    //    Then checking occurrences of 2.
    assert_eq!( test_bag.occurrences( 2 ), 2 );

    // G. Inserting the numbers 5, 6, and 7 into the bag. Then checking size.
    test_bag.insert( 5 );
    test_bag.insert( 6 );
    test_bag.insert( 7 );
    assert_eq!( test_bag.len(), 8 );

    // H. Inserting two more 2's into the bag. And then checking occurrences of 2's.
    test_bag.insert( 2 );
    test_bag.insert( 2 );
    assert_eq!( test_bag.occurrences( 2 ), 4 );

    // I. Inserting TEST_SIZE random items between 0 and 49 and then checking size.

    for _ in 0..TEST_SIZE
    {
        test_bag.insert( ( rand::random::<i16>() % 50 ) );
    }

    assert_eq!( test_bag.len(), TEST_SIZE + 10 );
}

#[ test ]
///   Performs some tests of the copy constructor and == method.
fn test2()
{
    let mut test_bag = Bag::new();

    // A. Testing that copy constructor works okay for empty bag...
    let copy1 = test_bag.clone();
    assert_eq!( copy1.len(), 0 );

    // B. Testing copy constructor with 4-item bag...
    test_bag.insert( 1 );
    test_bag.insert( 1 );
    test_bag.insert( 1 );
    test_bag.insert( 1 );
    let copy2 = test_bag.clone();

    //    and now testing the == method...
    assert_eq!( test_bag, copy2 );
    assert_eq!( copy2, test_bag );

    test_bag.insert( 1 ); // Alter the original, but not the copy
    // C. Then checking size of copy.
    assert_eq!( copy2.len(), 4 );

    // D. Altering original but not the copy.
    assert_eq!( test_bag.len(), 5 );
}

#[ test ]
///   Performs some tests of the assignment operator.
fn test3()
{
    let mut test_bag = Bag::new();
    let oldbytes;
    let newbytes;

    // A. Testing that assignment operator works okay for empty bag...
    let mut copy1 = Bag::new();
    copy1.insert( 1 );
    copy1 = test_bag.clone();

    assert_eq!( copy1.len(), 0 );

    // B. Testing assignment operator with 4-item bag...
    test_bag.insert( 1 );
    test_bag.insert( 1 );
    test_bag.insert( 1 );
    test_bag.insert( 1 );
    let mut copy2 = Bag::new();
    copy2.clone_from( &test_bag );
    test_bag.insert( 1 ); // Alter the original, but not the copy
    //    altering original by an insertion...

    assert_eq!( test_bag.occurrences( 1 ), 5 );
    assert_eq!( copy2.occurrences( 1 ), 4 );

    // Test passed.\n   testing size of assigned to...
    assert_eq!( copy2.len(), 4 );

    //    testing size of original...
    assert_eq!( test_bag.len(), 5 );

    // C. Testing assignment operator for a self-assignment...
    let mut hasher = ::std::collections::hash_map::DefaultHasher::new();
    oldbytes = ::std::hash::Hash::hash( &test_bag, &mut hasher );
    test_bag = test_bag;
    newbytes = ::std::hash::Hash::hash( &test_bag, &mut hasher );

    assert_eq!( oldbytes, newbytes );
}

#[ test ]
///   Performs basic tests for the erase functions.
fn test4()
{
    let mut test_bag = Bag::new();

    // Testing erase from empty bag, should have no effect.
    assert_eq!( test_bag.erase( 0 ), 0 );
    assert_eq!( test_bag.len(), 0 );
        
    // Inserting these: 8 6 10 1 7 10 15 3 13 2 5 11 14 4 12
    test_bag.insert(  8 );
    test_bag.insert(  6 );
    test_bag.insert( 10 );
    test_bag.insert(  1 );
    test_bag.insert(  7 );
    test_bag.insert( 10 );
    test_bag.insert( 15 );
    test_bag.insert(  3 );
    test_bag.insert( 13 );
    test_bag.insert(  2 );
    test_bag.insert(  5 );
    test_bag.insert( 11 );
    test_bag.insert( 14 );
    test_bag.insert(  4 );
    test_bag.insert( 12 );
    assert_eq!( test_bag.len(), 15 );

    // Now testing capacity, should be 16.
    assert_eq!( test_bag.capacity(), 16 );

    // Erasing 0, which is not in bag, so bag should be unchanged.
    assert!( !test_bag.erase_one( 0 ) );
    assert_eq!( test_bag.len(), 15 );

    // Erasing the 6.
    assert_eq!( test_bag.erase( 6 ), 1 );
    assert_eq!( test_bag.len(), 14 );

    // Erasing one 10.
    assert!( test_bag.erase_one( 10 ) );
    assert_eq!( test_bag.len(), 13 );

    // Erasing the 1.
    assert_eq!( test_bag.erase( 1 ), 1 );
    assert_eq!( test_bag.len(), 12 );

    // Erasing the 15.
    assert_eq!( test_bag.erase( 15 ), 1 );
    assert_eq!( test_bag.len(), 11 );

    // Erasing the 5.
    assert_eq!( test_bag.erase( 5 ), 1 );
    assert_eq!( test_bag.len(), 10 );

    // Erasing the 11.
    assert_eq!( test_bag.erase( 11 ), 1 );
    assert_eq!( test_bag.len(), 9 );

    // Erasing the 3.
    assert_eq!( test_bag.erase( 3 ), 1 );
    assert_eq!( test_bag.len(), 8 );

    // Erasing the 13.
    assert_eq!( test_bag.erase( 13 ), 1 );
    assert_eq!( test_bag.len(), 7 );

    // Erasing the 2.
    assert_eq!( test_bag.erase( 2 ), 1 );
    assert_eq!( test_bag.len(), 6 );

    // Erasing the one and only 14.
    assert!( test_bag.erase_one( 14 ) );
    assert_eq!( test_bag.len(), 5 );

    // Erasing the 4.
    assert_eq!( test_bag.erase( 4 ), 1 );
    assert_eq!( test_bag.len(), 4 );

    // Erasing the 12.
    assert_eq!( test_bag.erase( 12 ), 1 );
    assert_eq!( test_bag.len(), 3 );

    // Erasing the 8.
    assert_eq!( test_bag.erase( 8 ), 1 );
    assert_eq!( test_bag.len(), 2 );

    // Erasing the 7.
    assert_eq!( test_bag.erase( 7 ), 1 );
    assert_eq!( test_bag.len(), 1 );

    // Erasing the other 10.
    assert!( test_bag.erase_one( 10 ) );
    assert_eq!( test_bag.len(), 0 );

    // Testing capacity again.
    assert_eq!( test_bag.capacity(), 16 );

    // Now trimming to size.
    test_bag.shrink_to_fit();
    assert_eq!( test_bag.capacity(), 1 );

    // Now trimming to size again.
    test_bag.shrink_to_fit();
    assert_eq!( test_bag.capacity(), 1 );

    // Inserting value 5000 into the bag.
    // Inserting three 5's into the bag and then erasing all of them.
    test_bag.insert( 5000 );
    test_bag.insert( 5 ); 
    test_bag.insert( 5 );
    test_bag.insert( 5 );
    assert_eq!( test_bag.erase( 5 ), 3 );
    assert_eq!( test_bag.len(), 1 );
}

#[ test ]
///   Performs basic tests for the += and + functions
fn test5()
{
    let mut test_bag1 = Bag::new();
    let mut test_bag2 = Bag::new();
    let mut test_bag3 = Bag::new();

    // Inserting 2000 1's into test_bag1 and 2000 2's into test_bag2.

    for _ in 0..2000
    {
        test_bag1.insert( 1 );
        test_bag2.insert( 2 );
    }

    // Now testing the += function.
    test_bag1 += test_bag2.clone();

    //   and now testing for occurrences of 1's and 2's in test_bag1.
    assert_eq!( test_bag1.occurrences( 1 ), 2000 ); 
    assert_eq!( test_bag2.occurrences( 2 ), 2000 );

    // Now testing the + function.
    test_bag3.clone_from( &( test_bag2.clone() + test_bag2.clone() ) );

    //   and now testing for occurrences of 2's in test_bag3.
    assert_eq!( test_bag3.occurrences( 2 ), 4000 );
}

#[ test ]
///   Performs tests for the bag iterator
fn test6()
{
    let mut testbag = Bag::new();

    { // Testing next(), last(), and count()
        let mut iter = testbag.into_iter();
        assert_eq!( iter.next(), None );
        assert_eq!( iter.last(), None );
        assert_eq!( testbag.into_iter().count(), 0 );
    }

    testbag.insert( 8 );

    { // Testing next(), last(), and count() again
        let mut iter = testbag.into_iter();
        assert_eq!( iter.next(), Some( &8 ) );
        assert_eq!( iter.last(), Some( &8 ) );
        assert_eq!( testbag.into_iter().count(), 1 );
    }

    testbag.insert( 3 );
    testbag.insert( 4 );
    testbag.insert( 7 );

    { // Testing next(), last(), and count() again
        let mut iter = testbag.into_iter();
        assert_eq!( iter.next(), Some( &8 ) );
        assert_eq!( iter.last(), Some( &7 ) );
        assert_eq!( testbag.into_iter().count(), 4 );
    }

    testbag.insert( 2 );

    {
        let mut iter = testbag.into_iter();

        // Testing the full range of the bag with next()
        assert_eq!( iter.next(), Some( &8 ) );
        assert_eq!( iter.next(), Some( &3 ) );
        assert_eq!( iter.next(), Some( &4 ) );
        assert_eq!( iter.next(), Some( &7 ) );
        assert_eq!( iter.next(), Some( &2 ) );

        // ... and then None once it's over.
        assert_eq!( iter.next(), None );

        // More calls may or may not return None. Here, they always will.
        assert_eq!( iter.next(), None );
        assert_eq!( iter.next(), None );
    }
    { // Testing the nth( _ ) method
        let mut iter = testbag.into_iter();

        assert_eq!( iter.nth( 1 ), Some( &3 ) );
        assert_eq!( iter.nth( 1 ), Some( &7 ) );
        assert_eq!( iter.nth( 1 ), None );
    }
    {
        let mut iter = testbag.into_iter();

        assert_eq!( iter.nth( 0 ), Some( &8 ) );
        assert_eq!( iter.nth( 2 ), Some( &7 ) );
        assert_eq!( iter.nth( 9 ), None );
    }
}
